File: benchmark/05_mandelbrot/mandelbrot.go
// Benchmark 05: Mandelbrot Set
// Category: Floating Point / Simulation
// Tests: Complex arithmetic, nested loops, floating-point comparisons

package main

import (
	"fmt"
	"time"
)

const (
	WIDTH    = 1000
	HEIGHT   = 1000
	MAX_ITER = 256
)

func main() {
	start := time.Now()

	count := 0

	// Mandelbrot calculation
	for py := 0; py < HEIGHT; py++ {
		for px := 0; px < WIDTH; px++ {
			// Map pixel to complex plane
			x0 := float64(px-WIDTH/2) * 4.0 / float64(WIDTH)
			y0 := float64(py-HEIGHT/2) * 4.0 / float64(HEIGHT)

			x, y := 0.0, 0.0
			iter := 0

			for x*x+y*y <= 4.0 && iter < MAX_ITER {
				xTemp := x*x - y*y + x0
				y = 2.0*x*y + y0
				x = xTemp
				iter++
			}

			if iter == MAX_ITER {
				count++
			}
		}
	}

	elapsed := time.Since(start)

	fmt.Printf("Image size: %dx%d\n", WIDTH, HEIGHT)
	fmt.Printf("Points in Mandelbrot set: %d\n", count)
	fmt.Printf("Time: %.4f seconds\n", elapsed.Seconds())
}

File: benchmark/05_mandelbrot/mandelbrot.rs
// Benchmark 05: Mandelbrot Set
// Category: Floating Point / Simulation
// Tests: Complex arithmetic, nested loops, floating-point comparisons

use std::time::Instant;

const WIDTH: usize = 1000;
const HEIGHT: usize = 1000;
const MAX_ITER: usize = 256;

fn main() {
    let start = Instant::now();

    let mut count = 0;

    // Mandelbrot calculation
    for py in 0..HEIGHT {
        for px in 0..WIDTH {
            // Map pixel to complex plane
            let x0 = (px as f64 - WIDTH as f64 / 2.0) * 4.0 / WIDTH as f64;
            let y0 = (py as f64 - HEIGHT as f64 / 2.0) * 4.0 / HEIGHT as f64;

            let mut x = 0.0f64;
            let mut y = 0.0f64;
            let mut iter = 0;

            while x * x + y * y <= 4.0 && iter < MAX_ITER {
                let x_temp = x * x - y * y + x0;
                y = 2.0 * x * y + y0;
                x = x_temp;
                iter += 1;
            }

            if iter == MAX_ITER {
                count += 1;
            }
        }
    }

    let elapsed = start.elapsed();

    println!("Image size: {}x{}", WIDTH, HEIGHT);
    println!("Points in Mandelbrot set: {}", count);
    println!("Time: {:.4} seconds", elapsed.as_secs_f64());
}
